### PR TITLE
Chunked encoding for CCR APIs

### PR DESCRIPTION
**Pull request by DaveCTurner**
The CCR info and stats APIs can send fairly sizeable responses that scale as `O(#shards)`. This commit makes them chunked.

Relates #89838

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java [hunk 1]
```diff
@@ -32,7 +38,17 @@ public String getName() {
     @Override
     protected RestChannelConsumer prepareRequest(final RestRequest restRequest, final NodeClient client) {
         final CcrStatsAction.Request request = new CcrStatsAction.Request();
-        return channel -> client.execute(CcrStatsAction.INSTANCE, request, new RestToXContentListener<>(channel));
+        return channel -> client.execute(
+            CcrStatsAction.INSTANCE,
+            request,
+            new ThreadedActionListener<>(
```
Comment (original-brownbear) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java:44]: Why do I still need a threaded listener with chunked encoding?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestFollowStatsAction.java [hunk 2]
```diff
@@ -34,7 +40,17 @@ public String getName() {
     protected RestChannelConsumer prepareRequest(final RestRequest restRequest, final NodeClient client) {
         final FollowStatsAction.StatsRequest request = new FollowStatsAction.StatsRequest();
         request.setIndices(Strings.splitStringByCommaToArray(restRequest.param("index")));
-        return channel -> client.execute(FollowStatsAction.INSTANCE, request, new RestToXContentListener<>(channel));
+        return channel -> client.execute(
+            FollowStatsAction.INSTANCE,
+            request,
+            new ThreadedActionListener<>(
```
Comment (original-brownbear) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestFollowStatsAction.java:46]: Same here why does this need threading if it's chunked encoding?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/action/CcrStatsAction.java [hunk 3]
```diff
@@ -79,14 +81,14 @@ public void writeTo(StreamOutput out) throws IOException {
...
-            }
-            builder.endObject();
-            return builder;
+        public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params outerParams) {
+            return Iterators.concat(
+                Iterators.single(
+                    (builder, params) -> builder.startObject().field("auto_follow_stats", autoFollowStats, params).field("follow_stats")
+                ),
+                followStats.toXContentChunked(outerParams),
+                Iterators.single((builder, params) -> builder.endObject())
```
Comment (original-brownbear) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/action/CcrStatsAction.java:90]: NIT: I added `org.elasticsearch.common.xcontent.ChunkedToXContentHelper#endObject` for this.

**Review by original-brownbear — COMMENTED** [review]
Looks good just a random NIT and a question on the threading.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java (hunk 1 above)
Comment (DaveCTurner) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/rest/RestCcrStatsAction.java:44]: The `toXContent` implementation collects all the shard responses into a map keyed by index up-front - I think that should be done elsewhere.

https://github.com/elastic/elasticsearch/blob/c1efe471fb7d91e0275882bf7339bc2ce04aff66/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/action/FollowStatsAction.java#L73-L78

**Review by original-brownbear — APPROVED** [review]
LGTM thanks for the explanation.

**Pull request merged**